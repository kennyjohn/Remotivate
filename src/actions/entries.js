import database from '../firebase/firebase';

// Action Creators
// ADD_ENTRY
export const addEntry = (entry) => ({
  type: 'ADD_ENTRY',
  entry
});

export const startAddEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { 
      text = '',
      author = '',
      video = '',
      image = '',
      purpose = '',
      createdAt = 0
    } = entryData;

    const entry = { text, author, video, image, purpose, createdAt };
    return database.ref(`users/${uid}/entries`).push(entry)
    .then((ref) => {
      dispatch(addEntry({
        id: ref.key,
        ...entry
      }));
    });
  };
};

// REMOVE_ENTRY
export const removeEntry = ({ id } = {}) => ({
    type: 'REMOVE_ENTRY',
    id
});

export const startRemoveEntry = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).remove()
        .then(() => dispatch(removeEntry({ id })));
    };
};

// EDIT_ENTRY
export const editEntry = (id, updates) => ({
    type: 'EDIT_ENTRY',
    id,
    updates
});

export const startEditEntry = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).update(updates)
        .then(() => dispatch(editEntry(id, updates)))
    };
}

export const setEntries = (entries) => ({
  type: 'SET_ENTRIES',
  entries
});

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries`).once('value').then((snapshot) => {
      const entries = [];
      snapshot.forEach((childSnapshot) => {
        entries.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      dispatch(setEntries(entries));
      });
    });
  };
};