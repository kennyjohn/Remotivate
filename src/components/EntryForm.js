import React from 'react';
import moment from 'moment';

export default class EntryForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: props.entry ? props.entry.text : '',
            author: props.entry ? props.entry.author : '',
            video: props.entry ? props.entry.video : '',
            image: props.entry ? props.entry.image : '',
            purpose: props.entry ? props.entry.purpose : '',
            createdAt: props.entry ? moment(props.entry.createdAt) : moment(),
            disabledTextFlag: true, 
            disabledImageFlag: true, 
            disabledVideoFlag: true,
            error: ''
        };
    }  
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({ text }));
    }
    onAuthorChange = (e) => {
        const author = e.target.value;
        this.setState(() => ({ author }));
    }
    onVideoURLChange = (e) => {
        const url = e.target.value;
        this.setState(() => ({ video: url }));
    }
    onImageURLChange = (e) => {
        const url = e.target.value;
        this.setState(() => ({ image: url }));
    }
    onPurposeChange = (e) => {
        const purpose = e.target.value;
        this.setState(() => ({ purpose }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        const clicked = this.props.buttonClicked;
        if(!this.state.text && clicked === 'text') {
            this.setState(() => ({ error: 'Please provide text for your text entry.' }));
            return;
        }
        else if(!this.state.video && clicked === 'video') {
            this.setState(() => ({ error: 'Please provide a URL for your video entry.' }));
            return;
        }
        else if(!this.state.image && clicked === 'image') {
            this.setState(() => ({ error: 'Please provide a URL for your image entry.' }));
            return;
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                text: this.state.text,
                author: this.state.author,
                video: this.state.video,
                image: this.state.image,
                purpose: this.state.purpose,
                createdAt: this.state.createdAt.valueOf()
            });  
        }
    };
    componentDidMount() {
        if(this.state.text) {
            this.setState(() => ({
                disabledTextFlag: false,
                disabledImageFlag: true,
                disabledVideoFlag: true
            }));
        } else if(this.state.image) {
            this.setState(() => ({
                disabledTextFlag: true,
                disabledImageFlag: false,
                disabledVideoFlag: true
            }));
        } else {
            this.setState(() => ({
                disabledTextFlag: true,
                disabledImageFlag: true,
                disabledVideoFlag: false
            }));
        }
    }
    render() {
        return (
            <div>
                {this.props.buttonClicked && <h1 className="form__entry-header">Entry Form</h1>}
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <textarea
                    /* disabled attribute accepts single attribute boolean (e.g. true, false);
                    can't refer to a function to return boolean result */
                    maxLength="1000"
                    disabled={!this.props.buttonClicked ? this.state.disabledTextFlag : this.props.buttonClicked !== "text"}
                    placeholder="Text"
                    onChange={this.onTextChange}
                    value={this.state.text} 
                    className="textarea"
                    >
                    </textarea>
                    <input 
                    maxLength="70"
                    type="text"
                    placeholder="Author (Optional)"
                    onChange={this.onAuthorChange}
                    value={this.state.author}
                    className="text-input"
                    />
                    <input
                    disabled={!this.props.buttonClicked ? this.state.disabledVideoFlag : this.props.buttonClicked !== "video"}
                    type="text"
                    placeholder="Youtube Video URL"
                    onChange={this.onVideoURLChange}
                    value={this.state.video}
                    className="text-input"
                    />
                    <input
                    disabled={!this.props.buttonClicked ? this.state.disabledImageFlag : this.props.buttonClicked !== "image"}
                    type="text"
                    placeholder="Image URL"
                    onChange={this.onImageURLChange}
                    value={this.state.image}
                    className="text-input"
                    />
                    <textarea
                    placeholder={`Why is this ${!this.state.disabledTextFlag || this.props.buttonClicked === 'text' ? 'text' : 
                    !this.state.disabledImageFlag || this.props.buttonClicked === 'image' ? 'image' :
                    !this.state.disabledVideoFlag || this.props.buttonClicked === 'video' ? 'video' :
                    ''} important to you?`}
                    onChange={this.onPurposeChange}
                    value={this.state.purpose}
                    className="textarea"
                    >
                    </textarea>
                    <div>
                        <button className="button button--save">Save Entry</button>
                    </div>
                </form>
            </div>
        );
    }
}