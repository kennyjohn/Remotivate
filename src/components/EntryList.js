import React from 'react';
import { connect } from 'react-redux';
import EntryListItem from './EntryListItem';
import selectEntries from '../selectors/entries';

const EntryList = (props) => (
    <div className="content-container content-container--stretch list-grid">
         {
            props.entries.length === 0 ? (
            <div className="list-item list-item--message">
                <span>No entries at the moment</span>
                </div>
            ) : ( 
                props.entries.map((entry, index) => {
                    return <EntryListItem key={index} {...entry}/>;
                })
            )
        }
    </div>
  );

  const mapStateToProps = (state) => {
    return {
      entries: selectEntries(state.entries, state.filters)
    };
  };
  
  export default connect(mapStateToProps)(EntryList);