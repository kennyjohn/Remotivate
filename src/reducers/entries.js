const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_ENTRY':
            return [
                ...state,
                action.entry
            ];
        case 'REMOVE_ENTRY':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_ENTRY':
            return state.map((entry) => {
                if(entry.id === action.id) {
                    return [
                        ...state,
                        ...action.updates
                    ];
                } else {
                    return entry;
                }
            });
        case 'SET_ENTRIES':
            return action.entries;
        default:
            return state;
    }
}