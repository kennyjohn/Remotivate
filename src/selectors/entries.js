import moment from 'moment';

export default (entries, { value, sortByDate, startDate, endDate }) => {
    return entries.filter((entry) => {
        const createdAtMoment = moment(entry.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textFilterMatch = entry.text ? true : false;
        const videoFilterMatch = entry.video ? true : false;
        const imageFilterMatch = entry.image ? true : false;
        if(value === "text") {
            return startDateMatch && endDateMatch && textFilterMatch;
        } else if (value === "video") {
            return startDateMatch && endDateMatch && videoFilterMatch;
        } else if (value === "image") {
            return startDateMatch && endDateMatch && imageFilterMatch;
        } else {
            return startDateMatch && endDateMatch;
        }
    }).sort((a, b) => {
      if (sortByDate) {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    });
};
  