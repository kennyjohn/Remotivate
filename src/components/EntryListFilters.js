import React from 'react';
import { connect } from 'react-redux';
import { setFilter, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/lib/css/_datepicker.css';
import SelectEntryType from './SelectEntryType';
import { DateRangePicker } from 'react-dates';

export class EntryListFilters extends React.Component {
    state = {
        selectedFilter: null,
        calendarFocused: null
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };
    onFilterClick = (e) => {
        const value = e.target.value;
        /* unsets filter when button toggled  */
        if(this.state.selectedFilter === value) {
            this.props.setFilter(null);
            return this.setState({selectedFilter: null});
        }
        /* sets filter */
        this.props.setFilter(value);
        this.setState({selectedFilter: value});
    };
    componentDidMount() {
        /* selectedFilter should always be null upon component render */
        this.props.setFilter(this.state.selectedFilter);
    }
    render() { 
        const selectedFilter = this.state.selectedFilter;
        return (
            <div className="content-container content-container--stretch">
                <div className="input-group">
                    <span className="input-group__header">Filters:</span>
                    <div className="input-group__item">
                        <button 
                        onClick={this.onFilterClick}
                        value="text"
                        className={selectedFilter === 'text' ? 'button__filters--selected' : 'button__filters'}>
                        Text</button>
                        <button  
                        onClick={this.onFilterClick}
                        value="video"
                        className={selectedFilter === 'video' ? 'button__filters--selected' : 'button__filters'}>
                        Video</button>
                        <button  
                        onClick={this.onFilterClick}
                        value="image"
                        className={selectedFilter === 'image' ? 'button__filters--selected' : 'button__filters'}>
                        Image</button>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                        ></DateRangePicker>
                    </div>
                    <div className="input-group__item">
                       <SelectEntryType></SelectEntryType>
                    </div>
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setFilter: (value) => dispatch(setFilter(value)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(EntryListFilters);