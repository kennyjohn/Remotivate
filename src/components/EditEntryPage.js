import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import EntryForm from './EntryForm';
import { startEditEntry, startRemoveEntry } from '../actions/entries';

export class EditEntryPage extends React.Component {
    onSubmit = (entry) => {
        this.props.startEditEntry(this.props.entry.id, entry);
        window.location.reload(); /*Temporary fix*/
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveEntry({id: this.props.entry.id});
        this.props.history.push('/');
    }
    componentDidMount() {
        window.scrollTo(0, 0)
    }
    render() {
        let text = this.props.entry.text;
        let video = this.props.entry.video;
        let image = this.props.entry.image;
        let createdAt = this.props.entry.createdAt;
        return (
          <div>
            <div className="page-header">
              <div className="content-container">
                <h1 className="page-header__title">Edit Entry</h1>
              </div>
            </div>

           <div className={text ? "edit-entry__preview" : video || image ? "edit-entry__preview media-entry" : ""}>
                {   image ? <img src={image.replace("http:", "https:")}></img> :
                    text ? <p>{text}</p> :
                    video ? <iframe width="325" height="275" src={video.replace("watch?v=", "embed/")} frameBorder="0" allow="encrypted-media" allowFullScreen></iframe> :
                    null
                }
                <p className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</p>
           </div>

            <div className="content-container">
                <EntryForm
                  entry={this.props.entry}
                  onSubmit={this.onSubmit}
                />
                <button className="button button--remove button__extend" onClick={this.onRemove}>Remove Entry</button>
            </div>
        </div>
        );
      }
};

const mapStateToProps = (state, props) => ({
    entry: state.entries.find((entry) =>
    entry.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
    startEditEntry: (id, entry) => dispatch(startEditEntry(id, entry)),
    startRemoveEntry: (data) => dispatch(startRemoveEntry(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditEntryPage);