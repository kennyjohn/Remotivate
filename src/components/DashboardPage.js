import React from 'react';
import EntryList from './EntryList';
import Spotlight from './Spotlight';
import EntryListFilters from './EntryListFilters';

const DashboardPage = () => (
  <div>
    <Spotlight></Spotlight>
    <EntryListFilters></EntryListFilters>
    <EntryList></EntryList>
  </div>
);

export default DashboardPage;
