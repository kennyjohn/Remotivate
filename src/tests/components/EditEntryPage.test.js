import React from 'react';
import { shallow } from 'enzyme';
import entries from '../fixtures/entries';
import { EditEntryPage } from '../../components/EditEntryPage';

let startEditEntry, startRemoveEntry, history;
beforeEach(() => {
    /* Q: Why is beforeEach required instead of just defining the variables once? */
    startEditEntry = jest.fn();
    startRemoveEntry = jest.fn();
    history = { push: jest.fn() };
});

test('should render EditEntryPage component with the appropriate entry', () => {
    let wrapper = shallow(<EditEntryPage entry={entries[0]}></EditEntryPage>);
    expect(wrapper).toMatchSnapshot();
    wrapper = shallow(<EditEntryPage entry={entries[1]}></EditEntryPage>);
    expect(wrapper).toMatchSnapshot();
    wrapper = shallow(<EditEntryPage entry={entries[2]}></EditEntryPage>);
    expect(wrapper).toMatchSnapshot();
});

test('should handle EditEntry', () => {
    const wrapper = shallow(<EditEntryPage entry={entries[1]}
        startEditEntry={startEditEntry}
        startRemoveEntry={startRemoveEntry}
        history={history}
    ></EditEntryPage>);
    wrapper.find('EntryForm').prop('onSubmit')(entries[1]);
    expect(startEditEntry).toHaveBeenCalledWith(entries[1].id, entries[1]);
    /* The difference between toHaveBeenLastCalledWith and toHaveBeenCalledWith:   
    From my understanding, toHaveBeenCalledWith checks if the argument was called with
    history.push at all. For example, let's say you have a applyToAllFlavors(f) function
    that applies f to a bunch of flavors, and you want to ensure that the function was applied to 'apple'.
    The test case would still pass with toHaveBeenCalledWith, even if the last operation was applied to 'mango'. */
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle RemoveEntry', () => {
    const wrapper = shallow(<EditEntryPage entry={entries[1]}
        startEditEntry={startEditEntry}
        startRemoveEntry={startRemoveEntry}
        history={history}
    ></EditEntryPage>);
    wrapper.find('button').simulate('click');
    expect(startRemoveEntry).toHaveBeenCalledWith({ id: entries[1].id });
    expect(history.push).toHaveBeenLastCalledWith('/');
});