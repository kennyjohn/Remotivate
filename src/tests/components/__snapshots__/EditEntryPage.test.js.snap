// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditEntryPage component with the appropriate entry 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Entry
      </h1>
    </div>
  </div>
  <div
    className="edit-entry--preview"
  >
    <p>
      testing
    </p>
    <p
      className="list-item__subtitle"
    >
      December 31st, 1969
    </p>
  </div>
  <div
    className="content-container"
  >
    <EntryForm
      entry={
        Object {
          "author": "Kobe Bryant",
          "createdAt": 0,
          "id": "hqkwehk123",
          "image": "",
          "purpose": "",
          "text": "testing",
          "video": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--remove button__extend"
      onClick={[Function]}
    >
      Remove Entry
    </button>
  </div>
</div>
`;

exports[`should render EditEntryPage component with the appropriate entry 2`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Entry
      </h1>
    </div>
  </div>
  <div
    className="edit-entry--preview media-entry"
  >
    <img
      src="https://wwww.pictureimage.com"
    />
    <p
      className="list-item__subtitle"
    >
      January 5th, 1970
    </p>
  </div>
  <div
    className="content-container"
  >
    <EntryForm
      entry={
        Object {
          "author": "David Goggins",
          "createdAt": 432000000,
          "id": "69312332223",
          "image": "http://wwww.pictureimage.com",
          "purpose": "This means a lot to me because...",
          "text": "",
          "video": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--remove button__extend"
      onClick={[Function]}
    >
      Remove Entry
    </button>
  </div>
</div>
`;

exports[`should render EditEntryPage component with the appropriate entry 3`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Entry
      </h1>
    </div>
  </div>
  <div
    className="edit-entry--preview media-entry"
  >
    <iframe
      allow="encrypted-media"
      allowFullScreen={true}
      frameBorder="0"
      height="275"
      src="http://www.youtube.com"
      width="325"
    />
    <p
      className="list-item__subtitle"
    >
      December 26th, 1969
    </p>
  </div>
  <div
    className="content-container"
  >
    <EntryForm
      entry={
        Object {
          "author": "Jay Shetty",
          "createdAt": -432000000,
          "id": "wtfc00l",
          "image": "",
          "purpose": "",
          "text": "",
          "video": "http://www.youtube.com",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--remove button__extend"
      onClick={[Function]}
    >
      Remove Entry
    </button>
  </div>
</div>
`;
